### ISSUE TITLE
Pasting error messages into log panes sometimes doesn't create clickable links

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#2609]
The only workaround seems to be reloading Leo. The ekr-ann branch contains traces.

The new bpm-status command doesn't show anything amiss.

**Comment by tbpassin** [comment]
what  branch?  i'll see what i can do one-handed.

**Comment by tbpassin** [comment]
where is the code that tries to do the pasting?  Can you provide a sample message that failed?

**Comment by edreamleo** [comment]
@tbpassin Hey! Great to hear from you.

The branch: ekr-ann. The problem happens when pasting in the log pane, so the relevant code is: LeoFrame.pasteText.  The new special case is:
```python
    if wname.startswith('log'):
        if c.frame.log.put_html_links(s):
            return  # create_html_links has done all the work.
```
so the prime suspect is LeoLog.put_html_links.

The failures are mypy errors. For example:

`leo\plugins\qt_text.py:101: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]`

Reloading Leo (temporarily) fixes the problem, so I doubt that the actual error makes much difference.

LeoLog.put_html_links has a table of error regexs. leoGlobals.py defines these regexs in the section:

`<< define global error regexs >> (leoGlobals.py)`

**Comment by tbpassin** [comment]
can you induce the problem with messages that are not mypy errors?  maybe pylint errors?

if you stop getting the links, are the bare messages still getting printed or is there just nothing getting printed to the log?

**Comment by tbpassin** [comment]
ok, i was able to reproduce the error - no highlighted link in the pasted text.  i ran mypy to get some errors, copied one line from the console and pasted it into the log pane.  it pasted but no underlined clickable link.

not easy one-handed!

**Comment by edreamleo** [comment]
@tbpassin :-) Well done. For me the problem is intermittent. I'm glad you could reproduce it.

**Comment by tbpassin** [comment]
funny, i put a print statement at the start of the put_html_links statement and now i'm not getting the error.

    `print('====', s)`

btw, the mypy command is not limiting itself to the one file it's supposed to run against, but runs against many files.  pylint restrains itself just to the one file.

**Comment by edreamleo** [comment]
@tbpassin Especially when run from inside Leo, mypy seems to check to much. That's a big reason why I like to run mypy outside of Leo and then paste the messages into Leo's log pane.

**Comment by tbpassin** [comment]
it's checking too many files.  can't we graft on the file selection code from pylint ?

**Comment by edreamleo** [comment]
@tbpassin Interesting. Yes, it surely should be possible.

**Comment by edreamleo** [comment]
@tbpassin On the main issue here, there would seem to be a good quick check. When links do work,  LeoLog.put_html_links puts each line separately to the widget, even if a particular line doesn't match any regex. As a result, put_html_links is noticeably slower than pasting all at one in LeoFrame.pasteText.

But when links don't work, the paste goes quickly, which would indicate that put_html_links is returning False right away.  So I'm thinking that somehow the following lines aren't always doing what is expected:
```python
lines = g.splitLines(s)
# Step 1: return [True] if no lines match. This is an efficiency measure.
if not any(pat.match(line) for line in lines for pat in self.error_patterns):
    return False  # The [caller] must handle s.
```
Note. I've corrected two parts of comments above, marked in square brackets.

For sure a trace right before the code returns False will indicate that something is happening. I'll put in the trace right away.

**Comment by edreamleo** [comment]
@tbpassin A trace in mypy.run shows that mypy can produce more output when started on a particular file than when started from my script. Here is the script:
```
rem: See leo-editor/.mypy.ini for exclusions!
rem: FAST: python -m mypy %*
rem: sitecustomize.py must add leo.repo/mypy to sys.path for local mypy to work.
cls
python C:\leo.repo\mypy\mypy\__main__.py leo --debug-cache %*
```
Anyway, I'm not too concerned about this difference :-)

**Comment by edreamleo** [comment]
@tbpassin Wow. mypy's caching scheme is mysterious.

When I check qt_text.py from inside Leo, a trace shows that mypy.run is *only* calling mypy for qt_text.py, but changes to other files cause many new mypy warnings.  My workaround is to always use the script shown above, which somehow keeps the messages down.

**Comment by tbpassin** [comment]
1. i had noted the incorrect comments but decided to wait for another tome to mention them.
2. i also thought that the regexes weren't firing sometimes  - maybe because of garbage at the start or end of a line - but had not specifically checked yet.  seems like we are having similar thoughts here.

**Comment by tbpassin** [comment]
note also that put_html_links is indented under put_nl.  should have no effect but probably not intended.

**Comment by edreamleo** [comment]
@tbpassin Thanks for your comments.

- Yes, it seems we are both having similar thoughts.
- I have corrected the comments in the ekr-ann branch.
- put_html_links appears in the expected place in my version of leoSettings.leo.

**Comment by tbpassin** [comment]
interesting read - https://blog.zulip.com/2016/10/13/static-types-in-python-oh-mypy/

**Comment by tbpassin** [comment]
i just fetched the branch half an hour ago and it was indented in leopyref.

**Comment by tbpassin** [comment]
i'm thinking that if you really want to get serious about mypy in leo, give up on python 3.6.  too many features come in at 3.7.  see https://mypy.readthedocs.io/en/latest/runtime_troubles.html

**Comment by tbpassin** [comment]
what does this construct mean in the regex? 

(.+?)

the ? makes the preceding character optional but that char is a metacharacter already. so i don't understand.

**Comment by edreamleo** [comment]
@tbpassin Several comments:

- I wouldn't worry about leoPyRef.leo. It will be updated from my leoPy.leo when releasing 6.6.1.
- `(.+?)` specifies a non-greedy match. This is essential so that the pattern doesn't match too much!

From the [re documentation](https://docs.python.org/3/library/re.html#regular-expression-syntax):

*?, +?, ??

The `'*', '+', and '?'` qualifiers are all greedy; they match as much text as possible. Sometimes this behavior isn’t desired; if the `RE <.*>` is matched against '<a> b <c>', it will match the entire string, and not just '<a>'. Adding ? after the qualifier makes it perform the match in non-greedy or minimal fashion; as few characters as possible will be matched. Using the RE <.*?> will match only '<a>'.

**Comment by edreamleo** [comment]
@tbpassin It's just barely possible that the problem lies in a weird interaction between to-be-pasted text and regexes, but the evidence doesn't seem to support this hypothesis.

**Comment by edreamleo** [comment]
@tbpassin I've just reassigned this issue to 6.6.1, which means that I'll put any fix into 6.6.1 *provided* that the fix seems simple enough.  But no pressure: I'm happy to delay the fix into 6.6.2 if necessary :-)

**Comment by edreamleo** [comment]
@tbpassin I've just posted a link to the "static types in python" articles at leo-editor. It's excellent.

**Comment by edreamleo** [comment]
@tbpassin re runtime troubles: Ouch! This is only one reasons why #2604 must wait until Leo 6.6.2.

I would really rather not have the code test for python version for all imports of the typing module!

I am not real happy about requiring python 3.7. Python 3.8 was a dud, imo, so I would rather go straight to 3.9.

**Comment by tbpassin** [comment]
i forgot about non-greedy matches ... only used them a few times.

re versions, for type-checking i could agree but for actually running the code, seems we still have folks who want or need to stay with 3.6 for some time.

**Comment by tbpassin** [comment]
--follow-imports=skip reduces a lot of the time and variability running mypy.  sure it reduces checking, but might be worth it in earlier stages..

**Comment by tbpassin** [comment]
also see re mypy daemon.

**Comment by tbpassin** [comment]
in putting links, the code iterates through all lines and patterns.  if no match, quit;  if match, iterate through all lines and patterns *again*.  seems like we could do with only one double iteration here.

also, once a pattern matches, we could iterate through using just that pattern.  or are you thinking that a block of copied messages might come from more than than one pattern ?  i don't.

**Comment by tbpassin** [comment]
what if the pattern didn't start at the start of a line?  right now, you have to capture the start of a line exactly or it may not match.  if the leading "^" were removed there would be a little leeway for an inaccurate copy that inadvertently wrapped around the start.

**Comment by edreamleo** [comment]
@tbpassin Comments:

- I agree: we can "decree" that type checking will only happen with Python 3.9+, and only (say) with Qt6. No need to require users to upgrade. So that's (mostly) a non-issue. There may be some mypy-related complications in defining type constants when the version constraints aren't met. I'll investigate today.
- The "double" iteration through lines is not an issue. The first pass is extremely fast.
- Imo, the ^ and $ anchors are useful. Maybe even essential to prevent matching too much. I want to leave them in if at all possible. I don't think the anchors are the cause of the (hehe) "missing links."

**Comment by edreamleo** [comment]
@tbpassin Two more comments:
- Yes, `--follow-imports=skip` is worth considering, but in my experience mypy is fast enough without it.
- About your suggestion only to test one pattern in when putting links. This would have essentially zero impact on speed. To the first and second approximations, all of the slowdown in put_html_links comes from the line-by-line calls to LeoLog.put, *regardless* of what the argument to LeoLog.put is. None of the other code affects speed! I'm quite sure of this. In any case, I don't care about speed, I just want to see the links!

**Comment by tbpassin** [comment]
that's another thing i was wondering about - why we don't save up all the lines and give them to the log pane at one time.  setHtml() shouldn't care.

currently i'm not missing links so i'm just flailing around with ideas that i can't verify.

**Comment by tbpassin** [comment]
strings copied off console, both are found by the regexes running with ctrl-b:

```
import re

mypy_pat = re.compile(r'^(.+?):([0-9]+): (error|note): (.*)\s*$')
pylint_pat = re.compile(r'^(.*):\s*([0-9]+)[,:]\s*[0-9]+:.*?\(.*\)\s*$')


pylintline = r'C:\Tom\git\leo-editor\leo\core\leoKeys.py:2784:4: R0913: Too many arguments (9/6) (too-many-arguments)'

mypyline = r'C:\Tom\git\leo-editor\leo\core\leoKeys.py:1870: error: "KeyHandlerClass" has no attribute "singleLineCommandList"; maybe "defineSingleLineCommands"?  [attr-defined]'


m = mypy_pat.match(mypyline)
n = pylint_pat.match(pylintline)
g.es('mypy:', m and m.groups())
g.es('pylint:', n and n.groups())
```
both strings pasted into log successfully.

one way for it to fail: if the @files aren't found,  then no links.

btw, mypy ``really`` doesn't like leoKeys.py.

**Comment by tbpassin** [comment]
the "^" anchor does nothing useful because of the wildcard ``.+`` - i tried wrapping around part of the previous line and the match succeeded - but the wrapped  lead-in would be considered part of the filename ...

**Comment by edreamleo** [comment]
@tbpassin Now I'm less sure about what is happening. I just got missing links, but they appeared "instantaneously". How is that possible? I would have sworn that LeoQtLog.put would be slow! I just added a new trace in put_html_links that reports the number of lines if there *is* a match. Maybe that will help.

If I had to guess, it is that sometimes there is a problem with line endings, so only one line gets sent to put.

**Comment by edreamleo** [comment]
@tbpassin I'm pretty sure I know what is happening. Not sure why yet. I put a trace at the end of put_html_links that tells whether any patterns inside the loop match. That trace report True, even though no links were visible. The only non-mysterious way that could happen is if self.find_at_file_node(filename) returns None.  The original code says:
    `# An unusual case, but not worth a message.`
but I'll trace repr(filename). The next time this fires we should know why find_at_file_node failed.

**Comment by edreamleo** [comment]
@tbpassin Well, it seems my first guess was correct. I just got the following trace:
```
put_html_links MATCH 1  # that is, one line.
put_html_links FOUND MATCH True
```
but the log shows more than 30 lines!  So maybe we have to munge line endings??

**Comment by tbpassin** [comment]
i think it must be not finding the filenames.  what is the default for g.splitlines? keep nl or not?  maybe just strip() each line will do it?

**Comment by tbpassin** [comment]
g.splitlines(): """"Split s into lines, preserving the number of lines and
    the endings of all lines, including the last line."""

we don't want those line ends, i'm sure.

**Comment by vitalije** [comment]
Have you check if this happens in Linux too? I looked at `find_at_file_node` code and it seems to me that it makes all possible paths but in the local OS spelling and compares them with the result of `p.anyAtFileName()` which returns just a part of p.h which might be spelled in the Linux style (using '/' as separator).

**Comment by vitalije** [comment]
Here is a function that finds all filenames in the outline with the corresponding UNLs. It may be used just once before loop, and not on each pasted line.
```python
def find_all_filenames(c):
    res = []
    slash = re.compile(r'/|\\')
    fsep = os.path.sep
    def to_os_path(pth):
        return fsep.join(slash.split(pth))

    def atpath(v, pth):
        if v.h.startswith('@path '):
            return os.path.join(pth, v.h[6:].strip())
        a,sep,b = v.b.partition('@path ')
        if not sep: return pth
        if a == '' or a.endswith('\n'):
            ptharg = b.partition('\n')[0]
            return os.path.join(pth, to_os_path(ptharg))
        return pth
    def it(v, unl, pth):
        npth = atpath(v, pth)
        for i, ch in enumerate(v.children):
            nunl = f'{unl}-->{ch.h}'
            fname = ch.anyAtFileNodeName()
            if fname:
                yield os.path.join(npth, to_os_path(fname)), nunl
            yield from it(ch, nunl, npth)
    for a,unl in it(c.hiddenRootNode, f'unl://{c.mFileName}#', c.openDirectory):
        res.append([a, unl])
    return res
```

The result can be turned into a `dict` and then in loop you can use matched filename as a key to get the corresponding UNL. For example:
```python
def put_html_links(self, s: str) -> bool:
    """
    If *any* line is s contains a matches against known error patterns,
    then output *all* lines in s to the log, and return True.
    Otherwise, return False
    """
    c = self.c
    lines = g.splitLines(s)
    # Step 1: return s if no lines match. This is an efficiency measure.
    if not any(pat.match(line) for line in lines for pat in self.error_patterns):
        return False  # The user must handle s.
    # Step between 1 and 2: find all filenames and corresponding UNLs
    unls = dict(find_all_filenames(c))
    # Step 2: Output each line using log.put, with or without a nodeLink kwarg.
    for line in lines:
        for filename_i, line_number_i, pattern in self.link_table:
            m = pattern.match(line)
            if m:
                filename = m.group(filename_i)
                line_number = m.group(line_number_i)
                unl = unls.get(filename)
                if unl:
                    nodeLink = f'{unl}::-{line_number}' # Use global line.
                    self.put(line, nodeLink)
                else:
                    # An unusual case, but not worth a message.
                    self.put(line)
                break
        else:  # no match
            self.put(line)
    return True  # This method has completely handled s.
```
Hope this will help.

**Comment by tbpassin** [comment]
yes, it seemed clear that linux would have to be checked, especially if filenames are not getting matched because of line endings.  the situation is especially annoying to me because there are several other places where all or all matching at-file-names are searched for through the entire tree. it seems like there should be one reference implementation that does that job.

unfortunately i had stopped getting the failures so it's been hard to prove what's going on.

**Comment by vitalije** [comment]
In order to reproduce the issue (under the assumption that the problem is in Linux/Windows filename differences) is to have a single `@clean a/b.py` node with the code inside the body which produces an error when run. For example:
```python
raise ValueError('my dummy error')
```
Make a sub-folder `a` and save the outline on Windows machine. Then run `python a\b.py` to produce the error output, copy it and paste in the Log. If the problem is Windows/Linux paths then it won't create clickable link. 
Currently I am not able to test this on Windows.

**Comment by tbpassin** [comment]
i know how to test it, but unfortunately i didn't save the examples that caused some failures - i assumed it would be easy to produce some more.  i'll be working one-handed for some time so it's very awkward.

**Comment by vitalije** [comment]
I have just found out that python_pat sometimes doesn't recognize the traceback line. Python sometimes adds `',in <module>'` after the filename and line number. Perhapse this is why sometimes it works sometimes don't.

**Comment by vitalije** [comment]
For example pasting the following works for me:

```
Traceback (most recent call last):
  File "C:\Users\vital\Desktop\2609\demo\a.py", line 3
    raise ValueError("dummy error")
ValueError: dummy error
```
but pasting this doesn't:
```
Traceback (most recent call last):
  File "C:\Users\vital\Desktop\2609\demo\a.py", line 3, in <module>
    raise ValueError("dummy error")
ValueError: dummy error
```
note 'in <module>' at the end of traceback line.

**Comment by tbpassin** [comment]
a good catch, but we've seen a failure to create links with both pylint and mypy messages so it must be more than that.

**Comment by edreamleo** [comment]
@vitalije Thanks for your investigations. I'm sure we will get to the bottom of this soon.

**Comment by tbpassin** [comment]
if anyone has a reproducible test case, would you change the split method to `lines = s.split('\n')` in `put_html_links()`?  i don't see any good reason to keep the line ends and they can only cause trouble if not neutral.  i'm not getting errors for pasting  pylint and mypy message blocks but i don't have a sample that does cause errors.

**Comment by edreamleo** [comment]
@tbpassin Done at rev 44a0b88 in ekr-ann. This rev adds complete tracing for all branches of put_html_links.  Happy hunting.

**Comment by tbpassin** [comment]
in leoqtlog.put there is a comment
```
    # *Note*: For reasons that I don't fully understand,
    #         all lines sent to the log must now end in a newline.
    #
```

this isn't right. though.  like any other html, if you want a line break, you must put one in, whether ``<br>`` or some other standard html construct.

Example:

```
log = c.frame.log

text = '''line 1<br>
line 2<br>
line 3<br>
'''.split('\n')

for l in text:
    log.logWidget.insertHtml(l)
```

**Comment by edreamleo** [comment]
@tbpassin LeoQtLog.to_html adds the `<br>` tags.

**Comment by tbpassin** [comment]
found that sometime after my post.

**Comment by edreamleo** [comment]
@tbpassin @vitalije At last I have been able to catch the bug as it occurs.  Dumps after successful and failing pastes follow.

I'm pretty sure that the traces will provide important clues. It's not yet clear what causes the failures!

**Comment by edreamleo** [comment]
Here are the dumps for a successful paste:
```

Leo 6.6.1-devel, ekr-ann branch, build be58455349
2022-04-19 05:20:03 -0500
Python 3.9.6, PyQt version 6.2.4
Windows 10 AMD64 (build 10.0.22000) SP0
put_html_links At least one match found
37 lines:
[
    '',
    'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache',
    '',
    'C:\\leo.repo\\leo-editor>echo off',
    'leo\\core\\leoBeautify.py:22: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:30: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:60: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:87: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:107: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:130: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:163: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:193: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:221: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:241: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:248: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:279: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:286: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:295: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:316: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:333: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:336: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:385: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:394: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:405: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:425: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:435: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:469: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:482: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: error: Function is missing a return type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: note: Use "-> None" if function does not return a value',
    'leo\\core\\leoBeautify.py:499: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:502: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:545: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\plugins\\qt_frame.py:3149: error: Function is missing a type annotation  [no-untyped-def]',
    'Found 29 errors in 2 files (checked 74 source files)',
    '',
    'C:\\leo.repo\\leo-editor>'
]
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache'
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>echo off'
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file leoBeautify.py
put_html_links Found p: @file ../plugins/qt_frame.py
put_html_links No match in line: 'Found 29 errors in 2 files (checked 74 source files)'
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>'
put_html_links Found True matches
w.toPlainText:
[
    '',
    'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache',
    '',
    'C:\\leo.repo\\leo-editor>echo off',
    'leo\\core\\leoBeautify.py:22: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:30: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:60: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:87: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:107: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:130: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:163: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:193: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:221: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:241: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:248: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:279: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:286: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:295: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:316: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:333: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:336: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:385: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:394: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:405: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:425: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:435: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:469: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:482: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: error: Function is missing a return type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: note: Use "-> None" if function does not return a value',
    'leo\\core\\leoBeautify.py:499: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:502: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:545: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\plugins\\qt_frame.py:3149: error: Function is missing a type annotation  [no-untyped-def]',
    'Found 29 errors in 2 files (checked 74 source files)',
    '',
    'C:\\leo.repo\\leo-editor>',
    ''
]
w.toHtml:
[
    '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">\\x0a<html><head><meta name="qrichtext" content="1" /><meta charset="utf-8" /><style type="text/css">\\x0ap, li { white-space: pre-wrap; }\\x0a</style></head><body style=" font-family:\'DejaVu Sans Mono\',\'Inconsolata\',\'Droid Sans Mono\'; font-size:12pt; font-weight:400; font-style:normal;">\\x0a<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#fdffdd;"><br />',
    'C:\\leo.repo\\leo-editor&gt;python\\xa0C:\\leo.repo\\mypy\\mypy\\__main__.py\\xa0leo\\xa0--debug-cache<br />',
    '<br />',
    'C:\\leo.repo\\leo-editor&gt;echo\\xa0off<br />',
    '</span><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-22"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:22:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-30"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:30:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-60"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:60:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-87"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:87:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-107"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:107:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-130"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:130:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-163"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:163:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-193"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:193:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-221"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:221:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-241"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:241:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-248"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:248:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-279"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:279:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-286"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:286:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-295"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:295:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-316"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:316:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-333"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:333:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-336"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:336:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-385"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:385:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-394"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:394:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-405"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:405:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-425"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:425:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-435"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:435:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-469"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:469:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-482"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:482:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-485"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:485:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0return\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:485:\\xa0note:\\xa0Use\\xa0&quot;-&gt;\\xa0None&quot;\\xa0if\\xa0function\\xa0does\\xa0not\\xa0return\\xa0a\\xa0value<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-499"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:499:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-502"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:502:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Core classes--&gt;@file leoBeautify.py::-545"><span style=" text-decoration: underline; color:#fdffdd;">leo\\core\\leoBeautify.py:545:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><a href="unl:///C:/leo.repo/leo-editor/leo/core/leoPy.leo#Code--&gt;Qt gui--&gt;@file ../plugins/qt_frame.py::-3149"><span style=" text-decoration: underline; color:#fdffdd;">leo\\plugins\\qt_frame.py:3149:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    '</span></a><span style=" color:#fdffdd;">Found\\xa029\\xa0errors\\xa0in\\xa02\\xa0files\\xa0(checked\\xa074\\xa0source\\xa0files)<br />',
    '<br />',
    'C:\\leo.repo\\leo-editor&gt;<br />',
    '</span></p></body></html><br />'
]

```

**Comment by edreamleo** [comment]
Here are the dumps after a failed test. Two things stand out:
1. find_at_file_node failed to find any nodes.
2. The html looks very different from the "success" html.

This is quite mysterious!
```

Leo 6.6.1-devel, ekr-ann branch, build be58455349
2022-04-19 05:20:03 -0500
Python 3.9.6, PyQt version 6.2.4
Windows 10 AMD64 (build 10.0.22000) SP0
put_html_links No initial matches found
3 lines:
[
    'if trace:',
    '        g.printObj([dump(z) for z in lines], tag=f"{len(lines)} lines")',
    '        # g.printObj(lines, tag=f"{len(lines)} lines")'
]
put_html_links At least one match found
37 lines:
[
    '',
    'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache',
    '',
    'C:\\leo.repo\\leo-editor>echo off',
    'leo\\core\\leoBeautify.py:22: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:30: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:60: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:87: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:107: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:130: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:163: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:193: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:221: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:241: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:248: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:279: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:286: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:295: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:316: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:333: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:336: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:385: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:394: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:405: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:425: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:435: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:469: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:482: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: error: Function is missing a return type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: note: Use "-> None" if function does not return a value',
    'leo\\core\\leoBeautify.py:499: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:502: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:545: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\plugins\\qt_frame.py:3149: error: Function is missing a type annotation  [no-untyped-def]',
    'Found 29 errors in 2 files (checked 74 source files)',
    '',
    'C:\\leo.repo\\leo-editor>'
]
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache'
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>echo off'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:22: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:30: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:60: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:87: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:107: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:130: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:163: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:193: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:221: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:241: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:248: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:279: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:286: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:295: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:316: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:333: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:336: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:385: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:394: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:405: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:425: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:435: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:469: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:482: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:485: error: Function is missing a return type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:485: note: Use "-> None" if function does not return a value'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:499: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:502: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\core\\leoBeautify.py' line: 'leo\\core\\leoBeautify.py:545: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No p! filename: 'leo\\plugins\\qt_frame.py' line: 'leo\\plugins\\qt_frame.py:3149: error: Function is missing a type annotation  [no-untyped-def]'
put_html_links No match in line: 'Found 29 errors in 2 files (checked 74 source files)'
put_html_links No match in line: ''
put_html_links No match in line: 'C:\\leo.repo\\leo-editor>'
put_html_links Found True matches
w.toPlainText:
[
    '',
    'C:\\leo.repo\\leo-editor>python C:\\leo.repo\\mypy\\mypy\\__main__.py leo --debug-cache',
    '',
    'C:\\leo.repo\\leo-editor>echo off',
    'leo\\core\\leoBeautify.py:22: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:30: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:60: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:87: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:107: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:130: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:163: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:193: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:221: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:241: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:248: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:279: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:286: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:295: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:316: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:333: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:336: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:385: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:394: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:405: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:425: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:435: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:469: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:482: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: error: Function is missing a return type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:485: note: Use "-> None" if function does not return a value',
    'leo\\core\\leoBeautify.py:499: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:502: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\core\\leoBeautify.py:545: error: Function is missing a type annotation  [no-untyped-def]',
    'leo\\plugins\\qt_frame.py:3149: error: Function is missing a type annotation  [no-untyped-def]',
    'Found 29 errors in 2 files (checked 74 source files)',
    '',
    'C:\\leo.repo\\leo-editor>',
    ''
]
w.toHtml:
[
    '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">\\x0a<html><head><meta name="qrichtext" content="1" /><meta charset="utf-8" /><style type="text/css">\\x0ap, li { white-space: pre-wrap; }\\x0a</style></head><body style=" font-family:\'DejaVu Sans Mono\',\'Inconsolata\',\'Droid Sans Mono\'; font-size:12pt; font-weight:400; font-style:normal;">\\x0a<p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#fdffdd;"><br />',
    'C:\\leo.repo\\leo-editor&gt;python\\xa0C:\\leo.repo\\mypy\\mypy\\__main__.py\\xa0leo\\xa0--debug-cache<br />',
    '<br />',
    'C:\\leo.repo\\leo-editor&gt;echo\\xa0off<br />',
    'leo\\core\\leoBeautify.py:22:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:30:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:60:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:87:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:107:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:130:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:163:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:193:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:221:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:241:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:248:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:279:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:286:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:295:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:316:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:333:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:336:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:385:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:394:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:405:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:425:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:435:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:469:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:482:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:485:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0return\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:485:\\xa0note:\\xa0Use\\xa0&quot;-&gt;\\xa0None&quot;\\xa0if\\xa0function\\xa0does\\xa0not\\xa0return\\xa0a\\xa0value<br />',
    'leo\\core\\leoBeautify.py:499:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:502:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\core\\leoBeautify.py:545:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'leo\\plugins\\qt_frame.py:3149:\\xa0error:\\xa0Function\\xa0is\\xa0missing\\xa0a\\xa0type\\xa0annotation\\xa0\\xa0[no-untyped-def]<br />',
    'Found\\xa029\\xa0errors\\xa0in\\xa02\\xa0files\\xa0(checked\\xa074\\xa0source\\xa0files)<br />',
    '<br />',
    'C:\\leo.repo\\leo-editor&gt;<br />',
    '</span></p></body></html><br />'
]
```

**Comment by edreamleo** [comment]
A recent rev adds the 'dump-log' command (aka 'log-dump'), so now (Doh!) we don't have to clear the log in order to see it!

**Comment by tbpassin** [comment]
weird!  i never tried log-clear.  i always pasted with ctrl-v.

as a minor point, i suggest that the paste command always move to the end before actually inserting the new text.  there is a move to the end in the code but it doesn't come first, iirc.

**Comment by tbpassin** [comment]
when i invoke the 'paste-text' command using the minibuffer, it pastes into the body even if i had previously selected the log pane.

**Comment by edreamleo** [comment]
@tbpassin Thanks for the bug report. I have just created #2617.

**Comment by tbpassin** [comment]
run pylint.  get messages with links.  copy a message from the console, paste into log pane : get link.  select all log text with ctrl-a and delete with del key. repaste same message - now no link.  now rerun pylint and get messages with links.  but still no link when pasting the message.

**Comment by tbpassin** [comment]
repeat above sequence except only select a few lines to delete with shift-up-arrow.  delete the selected text.  now pasted line gets its link.  still true when all log text has been selected using shift-uparrow.

difference is deleting by ctrl-a vs shift-uparrow.

**Comment by tbpassin** [comment]
log pane key-handling issue: ctrl-uparrow prints "ctrl+up" in log pane, ctrl-downarrow prints "ctrl+down".  related? don't know.

**Comment by tbpassin** [comment]
another clue: get long pylint output, select all with ctrl-a, then deselect a few lines at bottom with shift-uparrow, then delete with delete key.  after this, pasted pylint message has link.

all the above data obtained by running pylint on qt_frame.py, mostly after a restart.

**Comment by tbpassin** [comment]
another anomaly - select entire log text with ctrl-a.  then type some char, say "c".  any editor would delete all text, but log pane deselects and adds the typed char to previous text.  pasted pylint message has link after this.

**Comment by tbpassin** [comment]
Some of above observations may be conditional on something unknown, because after resetting all changes to the branch, I am able to paste console line with links after deleting pylint log output with ctrl-a / delete.  So there is still some other factor.

maybe something stateful?  a module-level variable that should be a class-level one?

**Comment by tbpassin** [comment]
here is another case that caused link paste failure.  after a restart, and pasting a link worked whether or not i had selected all log text and deleted or not, i opened a new tab by showing bindings. after this, pasting my example link into the log pane did not create the intended links. 

however, it is not simply creating another tab, because when i created my "vr4" tab, it did not interfere with pasting the link message into the log pane.

**Comment by tbpassin** [comment]
also, after some of these malfunctions, the up/down arrow keys do not work in the log pane, although they do in the newly added tab.

i am coming to think that sometimes the log pane widget is getting replaced by another one that does not go through the intended connect process.

**Comment by tbpassin** [comment]
the link paste failures after showing settings/bindings/commands repoed on linux mint.

both paste failure after ctrl-a / del and arrow key failure happened.

**Comment by edreamleo** [comment]
@tbpassin Thanks for all these tests :-) One aspect of link "failures" seems to be that the log pane clears key bindings.

My plan: have put_html_links (or some other code) warn if bindings no longer exist.

**Comment by edreamleo** [comment]
@vitalije Somewhere recently you posted a replacement for find_at_file_node. Better tracing has highlighted this method as not working. Could you please post your code here, or somewhere else that is convenient for you? Thanks.

For now, I'm going to try something a bit too simple :-)

**Comment by tbpassin** [comment]
for toggling the tab on and off for my tabbed version of vr3, i ended up caching the widget in the log frame's dict.  i restored from there.  otherwise the widget would get trashed when deactvated.
```
imports_ok = False
log = c.frame.log
TABNAME = 'VR4'

if log.contentsDict.get('vr4-visible', False):
    # Toggle the tab visibility
    w = log.contentsDict['vr4-widget']
    w.deactivate()  # Disconnects event handlers
    w.deleteLater()
    log.deleteTab(TABNAME)
    log.contentsDict['vr4-visible'] = False
else:
    if log.contentsDict.get('vr4-loaded', False):
        log.selectTab(TABNAME, widget = log.contentsDict['vr4-widget'])
        log.contentsDict['vr4-visible'] = True
    else:
        try:
            from vr4.viewrendered4 import ViewRenderedController4
            vr4_widget = ViewRenderedController4
            imports_ok = True
        except ImportError:
            g.es('Error importing VR4', color = 'red')

        if imports_ok:
            w = vr4_widget(c)
            log.selectTab(TABNAME, widget = w)
            log.contentsDict['vr4-loaded'] = True
            log.contentsDict['vr4-visible'] = True
            log.contentsDict['vr4-widget'] = w
            w.activate()  # Connects event handlers
```

**Comment by tbpassin** [comment]
i don't think the problem is with the code identifying the at-files.  i think that code isn't getting called because the log widget no longer has its bindings when the link line is pasted in.  we're probably seeing default behavior of a QTextBrowser widget.

**Comment by vitalije** [comment]
@edreamleo  here follows the message with code:

Here is a function that finds all filenames in the outline with the corresponding UNLs. It may be used just once before loop, and not on each pasted line.
```python
def find_all_filenames(c):
    res = []
    slash = re.compile(r'/|\\')
    fsep = os.path.sep
    def to_os_path(pth):
        return fsep.join(slash.split(pth))

    def atpath(v, pth):
        if v.h.startswith('@path '):
            return os.path.join(pth, v.h[6:].strip())
        a,sep,b = v.b.partition('@path ')
        if not sep: return pth
        if a == '' or a.endswith('\n'):
            ptharg = b.partition('\n')[0]
            return os.path.join(pth, to_os_path(ptharg))
        return pth
    def it(v, unl, pth):
        npth = atpath(v, pth)
        for i, ch in enumerate(v.children):
            nunl = f'{unl}-->{ch.h}'
            fname = ch.anyAtFileNodeName()
            if fname:
                yield os.path.join(npth, to_os_path(fname)), nunl
                # we don't need to search inside any of the file subtrees
            else: 
                yield from it(ch, nunl, npth)
    for a,unl in it(c.hiddenRootNode, f'unl://{c.mFileName}#', c.openDirectory):
        res.append([a, unl])
    return res
```

The result can be turned into a `dict` and then in loop you can use matched filename as a key to get the corresponding UNL. For example:
```python
def put_html_links(self, s: str) -> bool:
    """
    If *any* line is s contains a matches against known error patterns,
    then output *all* lines in s to the log, and return True.
    Otherwise, return False
    """
    c = self.c
    lines = g.splitLines(s)
    # Step 1: return s if no lines match. This is an efficiency measure.
    if not any(pat.match(line) for line in lines for pat in self.error_patterns):
        return False  # The user must handle s.
    # Step between 1 and 2: find all filenames and corresponding UNLs
    unls = dict(find_all_filenames(c))
    def getunl(fname):
        xs = [fname] if fname in unls else [x for x in unls if x.endswith(fname)]
        return unls[xs[0]]
    # Step 2: Output each line using log.put, with or without a nodeLink kwarg.
    for line in lines:
        for filename_i, line_number_i, pattern in self.link_table:
            m = pattern.match(line)
            if m:
                filename = m.group(filename_i)
                line_number = m.group(line_number_i)
                unl = getunl(filename)
                if unl:
                    nodeLink = f'{unl}::-{line_number}' # Use global line.
                    self.put(line, nodeLink)
                else:
                    # An unusual case, but not worth a message.
                    self.put(line)
                break
        else:  # no match
            self.put(line)
    return True  # This method has completely handled s.
```

If we can assume that all filenames in error messages are absolute, then there would be no need for the function `getunl(fname)` and instead `unls.get(filename)`  can be used.

**Comment by edreamleo** [comment]
@vitalije Thanks.

**Comment by vitalije** [comment]
I also think that the regular expressions used to extract the filename and line number are unnecessarily strict. They try to match the whole line, but all we need is just the filename and the line number. If they match some text that looks like a combination of a file and a line number, most likely this file won't exist in the outline so false positives are harmless. And if they really do match some node, I don't think user would mind.

Also, if `re.M` flag is added in the `re.compile` call, then it isn't necessary to search line by line to decide whether there will be any match or not. It would be faster to apply the pattern to the entire text and if it succeeds, then there must be at least one match.

**Comment by edreamleo** [comment]
@vitalije Thanks for these comments. Happily, we don't need to worry too much about edge cases.

I have just merged (by hand), the latest version of put_html_links into devel, so everyone can test it. For reasons that I do not entirely understand, put_html_links suddenly seems more robust than before.

Inspired by your code,put_html_links precomputes the **at_file_nodes** list. I've retained the **find_at_file_node** function. Imo, it is simpler, and possibly more robust, than the code you suggested.

As a completely separate matter, the **find_match** function simplifies put_html_links.

**Comment by edreamleo** [comment]
Rev a95f59 merges improved code from the ekr-ann branch into devel. The new code appears to work.

Please open a **new** issues if you discover more problems.

**Comment by vitalije** [comment]
@edreamleo I don't wish to argue too much about this, but my function does a lot less work than what your code does.

Let me point out just a few points:
1. `at_file_nodes = [p for p in c.all_positions() if p.isAnyAtFileNode()]` actually does full outline scan.
2. for each pasted line that match regexp, `at_file_nodes` is processed and for each item every time calcualting `os.path.normpath` several times depending on how much parts there are in the target absolute path.
3. when the position is found, then it is calculating the UNL for that position. 

while OTOH, my code, calculates the list of pairs (absolute file name, unl) much faster and it is calculated only once. UNLs are calculated on the fly during the outline traversal. In other words there are no extra work required.

I know, that the speed improvement of a (let's say) 100ms in this case,  is not much, but perhaps this function (finding quickly all absolute paths present in the outline), might be useful elsewhere in Leo core.

This is just a suggestion.

**Comment by edreamleo** [comment]
@vitalije I tried your code and it kept crashing. I'm open to a PR, but let's wait until the 6.6.2 gets merged into devel.

**Comment by edreamleo** [comment]
@tbpassin Thanks for this report. We can investigate after releasing 6.6.1.

**Comment by tbpassin** [comment]
sure. and you may be the only person who wants to use the log pane this way ...

**Comment by edreamleo** [comment]
@tbpassin Exactly. We are talking about edge cases in a little-used feature.

**Comment by edreamleo** [comment]
@vitalije On second thought, I would rather you not bothered with a PR. I can understand the present code and the present code is good enough.

**Comment by edreamleo** [comment]
@tbpassin @vitalije I am going to close this issue now.

If we discover *serious* problems, we can open a *new* issue for 6.6.1.

For less serious problems, let's create a new issue for 6.6.2.

**Comment by edreamleo** [comment]
FYI: Simple tests pass on Ubuntu.
